### ISSUE TITLE
Improvement: change counters to unsigned

### ISSUE THREAD

**Issue by 0bi-w6n-K3nobi** [allinurl/goaccess#1988]
Good morning, @allinurl .

Nice to be here again.
I've been too busy moving house.

Yes. The statistics I really need are intense and in high volume.
Unfortunately, when counting 1 month the counters exceeded 31 bits.
I easily checked the exact values in JSON, and replaced it by applying the complement of 2.
Anyway I am registering here.

![image](https://user-images.githubusercontent.com/71027865/103349945-2b22cb00-4a7d-11eb-86b7-823f18302b29.png) ![image](https://user-images.githubusercontent.com/71027865/103349924-121a1a00-4a7d-11eb-9996-870ccde42695.png)

![image](https://user-images.githubusercontent.com/71027865/103350049-79d06500-4a7d-11eb-89cf-91236d39588c.png)
![image](https://user-images.githubusercontent.com/71027865/103350095-94a2d980-4a7d-11eb-87e5-2da892622f66.png)

Perhaps an improvement would be to switch some counters to 64bit.

Thankful,
See you around.

**Comment by allinurl** [comment]
Thanks for reporting this. Less than 48hrs ago I was thinking about this exact same thing, certainly a coincidence! I'll start looking into it. I think all the plumbing is done so it shouldn't be a major change to flip the switch to a 64bit. I wonder if it's worth adding this as configure flag, e.g., `--with-64bit` so we don't necessarily increase the memory space for those who don't needed or even for all.

**Comment by allinurl** [comment]
BTW, could you please do a `wc -l` on all those logs? I just want to double check how many lines you are parsing. Thanks

**Comment by 0bi-w6n-K3nobi** [comment]
Good night, @allinurl.

I don't believe in coincidence much. But perhaps  to be in line/conformity with the  work.
Several times here I commented on something, that shortly after you implement it in the code.

Yes, I checked the line count.
As I mentioned, I totaled 1 month for **all sites**.
That is why these values are so high.

I believe that it would not be good to use all variables with 64 bits.
In the case mentioned above, I believe that used more than 16 Gigas to process the entire month.
I have available only 1 machine with 32 Gigabytes. And so, it would make it impossible to do the same processing entirely in 64 bits.
I believe it is enough to have 64 bits on the panels that have more grouping and less items. That is:
* Operating Systems;
* Browsers;
* HTTP Status Codes;
* Geo Location;
* MIME Types;
* Encryption settings;

And sure enough, all the totalizers in the panels/tables.

Well ... If you really want to save memory you should use a fixed-point/integer. That is:
2 bytes  -- 0xxx
4 bytes -- 10xxxx
8 bytes -- 11xxxxxx
But I believe that it can be complicated to implement.

To finish...
Happy New Year.
And that we can share/collaborate on this and other projects in the future.

See you.

**Comment by 0bi-w6n-K3nobi** [comment]
**PS:**
You could implement memory savings just by using a flag to tell whether it is 16, 32 or 64 bits in each data array.

**Comment by 0bi-w6n-K3nobi** [comment]
Good afternoon @allinurl.

Yes. It is probably also a good option to compile with the `--with-64bit` flag.
However it is necessary to create the **warning** that this would be necessary for **large volumes** of requests.
And that the memory consumption would be practically **doubled**.

In my particular case, I had a total of 2.5 billion requests with about 40 Terabytes of bandwidth.
Only I had problems with the `HTTP Status Codes`, `The cache status of the object served` and `Geo Location` panels.
And of course: total hits and total requests.
If the values were `UINT32_T` it would be limited to a maximum of 2 months.
Which would be good enough. But it would _not be_ possible to total _a year_.

See you.

**Comment by allinurl** [comment]
Hello @0bi-w6n-K3nobi. Thanks for the additional feedback. I'd need to double check on this cause it should be set to `uint32_t`. More likely I missed an int somewhere... I'll post back as soon as I have some news.

Also, I'll be looking at the 64 bit thing pretty soon and see how to best address it.

**Comment by 0bi-w6n-K3nobi** [comment]
Good evenin @allinurl.

So ... I think we could talk more about it here.
Perhaps a brain-storm about this.

First of all I would like to make some considerations about `GoAccess`:
* Bits == order of magnitude that really needed;
* Compiled in 64bits vs 32bits;

**Order of magnitude**

We can use the above mentioned values ​​as an example.
That is: 
2.5 Billion Hits with ~40 TiBytes of Tx Amount.
If I happened to want to count 1 year, I would have:
30.2 Billion Hits == ~35 Bits with ~460 TiBytes == ~49 Bits.
Under these circumstances, the processing time for 1 month would be 90 hours 
and for 1 years more than 1000 hours.

Now, let's think in reverse order: 
I believe that nobody can have a server larger than 256 GigaBytes of RAM == 38 Bits.
Even if We have a state-of-the-art Intel I7, processing time would not be feasible to occupy all that memory.
I have gotten a factor of 8 at least, which would be about 32 Billion of requests.
Well ... these are the values ​​in my example for 1 year.

In summary:
We are unable to use more than 35 bits for counters in general ... of course the exception is Tx Amount.

**Compiled at 64bits vs 32bits**

It is important to make some notes here.
All pointers are 64 bits when `GoAccess` is compiled at 64 bits.
In general I think this is not a problem.
However, the data structures for the panels make abundant use of pointers.
And as we saw above, we need a maximum of 38 bits to address the memory we can use.
What to do then?

Some options see in my mind:
* Use 48-bit pseudo-pointer;
* Use 32-bit pseudo-pointer;

How to implement?

In 48 bits:
Here I find it simple.
It just saves an offset for the addresses obtained via `malloc`.
And just subtract for generate pseudo-pointer or a mask reducing the 64 bits to 48.
It can also be fixed value, depending on the `OS` and use the `MACRO` feature.
Of course, use `bit` to represent the pseudo-pointer.

In 32 bits:
It is more elaborate here.
Maybe `HASH` or some table for reduction and represent the pointer in only 32 bits.

**Final notes**

Well ... All of this is related to `GoAccess` compiled in 64 bits.
If it is 32, just define the `MACROs` to bypass the values.

Another care must be taken, due to the pad made to align the members in the structures.
In Intel architecture, 32-bit/4-byte alignment is generally used.
On architectures like `SPARC` and `PowerPC`, this alignment can jump to 8 bytes if You have a `float` or `double` in the structure.
This is due to the implementation of the `FPU` unit in the `RISC` architecture.

Okay. I hope you're not reinventing the wheel.
And that this topic will be useful to you in something.

See you around.

**Comment by 0bi-w6n-K3nobi** [comment]
Hi @allinurl .

I find [this article](https://llvm.org/pubs/2005-06-12-MSP-PointerComp.pdf)... It is some complex but I believe than can help with some idea.

See you.

**Comment by 0bi-w6n-K3nobi** [comment]
Hello @allinurl.

I'm sorry for the previous post. I was completely wrong.
In order to know exactly the difference between 32 and 64 bits, I ran the program several times and built the table below:

```
======================================
requests     runtime [seconds]
[millions]
      32 bits  64 bits  diff    diff %
======================================
  0      0.3      0.2     0.1   -39.2
 10    810.8    656.7   154.1   -19.0
 20   1632.2   1333.9   298.3   -18.3
 30   2262.3   1806.5   455.9   -20.1
 40   2881.6   2335.8   545.8   -18.9
 50   3459.5   2847.2   612.3   -17.7
```

```
======================================
requests     memory [megabytes]
[millions]
      32 bits  64 bits  diff    diff %
======================================
  0      3.3      4.0     0.7   18.4
 10    587.6    632.4    44.9    7.1
 20   1300.4   1407.4   107.1    7.6
 30   1995.9   2171.6   175.7    8.1
 40   2709.7   2940.6   230.9    7.9
 50   3427.0   3728.1   301.1    8.1 
```
And also the following graphics:

for Runtime
![image](https://user-images.githubusercontent.com/71027865/104660483-69d39a80-56a5-11eb-96b9-7180a11ed4d4.png)

for Memory
![image](https://user-images.githubusercontent.com/71027865/104660595-a2737400-56a5-11eb-822b-04afb925a7b0.png)

Where the `orange line` means the `difference`. In the first, it was multiplied by 3.5 and in the second by 10 to fit the same scale.

If you do the linear regression for the `memory graph`, by the `difference`, you will see that 
the constant `a`, where `y = ax + b`, is between **6** to **7**.
So, _this is_ the value in _bytes per request_, on average, that the 64bit version **_overhead_** about to 32bit.
And again doing the regression again, over _64bit values_, we have ~ **83** --  i.e. the average _bytes per request_.

Phew ... Now yes, We have a correct statement about 32 and 64 bits.
Well ... I think I managed to clear up the misunderstanding.

Now, addressing the fact that we need to increase the number of bits for the counters -- i.e. the counters for each item -- 
I still find it appropriate to use `40 bits` [5 bytes] instead of the native `UINT64` type.
This would reduce memory usage well.
However it is necessary to be careful with the `alignment` and `pad` in the data structures.

Well ... These are my considerations on this subject.

See You.

**Comment by allinurl** [comment]
@0bi-w6n-K3nobi Thanks a lot for all these findings and suggestions. Pretty interesting charts that you posted there. I'm still reading and taking some ideas from the article you shared though. To be honest, I like the idea of using 40 bits while properly taking care of the alignment. I'm still thinking a bit more how best to approach this.

On the other hand I feel like we could approach large datasets (i.e., 4 billion+ requests) using an existing technology that could scale better. I'm working on an implementation of goaccess reading data from postgres. Note that I say reading since a different process will be taking care of the ingestion while goaccess will take care of the processing and outputting. This approach could technically be expanded to other dbs with minor implementations on the goaccess' end. This is still on the works so I can really comment much about it until I have something more tangible :)

**Comment by 0bi-w6n-K3nobi** [comment]
Good afternoon, @allinurl.

Yes. I think the DB implementation would also be nice.
However, I would like to make a note about this: I chose `GoAccess` because I don't need a database server.
I think the DB would be an add, though it shouldn't be mandatory.

Local database would also be nice: `MongoDB`, `NOSql`, etc.
Have you ever looked in the Gnome interface for data? - GDA or [GNOME Data Access](https://www.gnome-db.org/).
Although a little complex, as far as I can remember, it allowed using `XML` as a `DB`. In the current implementation it uses `SQLite`.
Another interesting project is [SQLRelay](http://sqlrelay.sourceforge.net/).

About the article I posted: there is a reference, # 1, to a JAVA implementation in which the programmer reduced the addressing to 32bits with performance gain. It is also worth reading. It's in [here](https://suif.stanford.edu/~brm/papers/10.1.1.97.8725.pdf). 

Well ... as for the 40 bits ... it is as simple as it looks.
Use `bit` type and do all intermediate operations in `UINT64_T`.
I believe that Language C will do the necessary conversions without problems, if obviously we do not exceed 40 bits.

So, until next time.

**Comment by allinurl** [comment]
@0bi-w6n-K3nobi I agree, the db would be an additional feature. One of the main concepts of goaccess has always been minimal dependencies and thus so far only needs ncurses and I'll make sure to keep it that way. I'm positive an implementation that supports 4B+ requests will make it upstream, just need to find a good way of doing it. I'll look further into the notes you posted. Thanks again!

**Comment by 0bi-w6n-K3nobi** [comment]
Hello @allinurl.

I think I understand your dilemma. And correct me if I'm wrong.
The problem is not just counters larger than 32 bits.
If we process LOGs with a volume greater than 4 Bi, we will also have a large memory consumption.
How could we solve this?

If we have a different input feature, not only a physical LOG, but also a DB, SolR, MemCache or Redis would be incredible.
But that does not address our memory consumption problem.
This is something that you will have to "meditate" on if you also want to process big data like 4Bi of requests.

Well ... This may be another problem, or another approach -- big data.
And that requires another method/approach to solve.
Maybe something like a local bank, just like you do with HASH tables, but on disk to minimize memory consumption.
To tell you the truth, I don't know what the best solution would be.

We can see how SolR, ElasticSearch and others solved the problem of large amounts of information.
Maybe that can give us an idea of ​​how to do it.

Well ... As you said, this is just a rough draft on the subject.

See you here.

**Comment by 0bi-w6n-K3nobi** [comment]
Good morning @allinurl.

Here, I will complement the subject of the last post.

Well ... I've had to work a lot with local key-value data.
At the time, I used the elder `GDBM` which was fast for reading but very bad for writing.
Later, I also used the same feature in the Python Language that is transparent to use in memory or on disk.

Why am I talking about this here?
I know the work you must have done to convert `Kyoto Cabinet's` `GoAccess` to its own implementation in `HASH tables`.
I do not underestimate this.

However, if we are going to address a very large volume of data we will have to think of a solution.
At the moment we talked, I am processing the month of December for my websites which I estimate as:
**2,5 Bi** of requests using **25 GiB** of memory.
And as I mentioned earlier, I only have 1 host with 32 GiB, a few others with 16 GiB and the rest with only 8 GiB of memory.

A `GoAccess` implementation with local key-value data on disk could be the solution for those with little memory, or 
who have to process absurd amounts of data.
Another option would be some process for using the memory of more than one host -- something like a key-value database in clustered memory. 
Something like `Redis`, `MemCache` on several hosts and `GoAccess` doing partitioning.
Maybe we went on to create 2 separate compiled versions -- on disk and in memory.
I know this contrasts with the philosophy that GoAccess wants to stay on -- simple, fast == everything in memory, and efficient.

I found two good implementations here: [LevelDB](https://github.com/google/leveldb) by Google and [LMDB](https://github.com/LMDB/lmdb/tree/mdb.master/libraries/liblmdb) by OpenLDAP. 
And Wikipedia references [here](https://en.wikipedia.org/wiki/LevelDB) and [there](https://en.wikipedia.org/wiki/Lightning_Memory-Mapped_Database).
Also a little old review [here](http://www.lmdb.tech/bench/microbench/).

Well ... It's still something to think about ... just an outline.

**Comment by 0bi-w6n-K3nobi** [comment]
Other good reference:

**Comment by 0bi-w6n-K3nobi** [comment]
More... [here](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.695.1842&rep=rep1&type=pdf).

**Comment by allinurl** [comment]
@0bi-w6n-K3nobi Thanks for those references. I'll be looking at those.

Just wanted to give you an update and let you know that https://github.com/allinurl/goaccess/commit/d7318ac9122f91b68188a1795f56c4ecf1e911f1 should let you output up until uint32_t (i.e., 4294967295) as it was originally intended. Please feel free to test this out by building from development and let me know how it goes. I'll be pushing out this in v1.4.4 pretty soon. Thanks.

**Comment by 0bi-w6n-K3nobi** [comment]
Hi @allinurl.

Thanks, I will test soon possible.

More one key-value db: [mdbx](https://github.com/erthink/libmdbx) and info [here](https://erthink.github.io/libmdbx/index.html).

**Comment by 0bi-w6n-K3nobi** [comment]
@allinurl ... Cool ... I took a look ... You changed the main totalizers. I believe that everything should be ok.

**Comment by 0bi-w6n-K3nobi** [comment]
Good afternoon, @allinurl.

I am very inclined to `LMDB`.
He has these articles/projects also about it:
[LMDB in Node](https://dev.doctorevidence.com/lmdb-in-node-29af907aad6e)
[LumoSQL](https://github.com/LumoSQL/LumoSQL) with [video](https://www.youtube.com/watch?v=rKtDeMvzZ2U)
[The Glitter and Poverty of LMDB Key-Value Database in iOS Apps](https://habr.com/ru/company/mailru/blog/480850/)

It would basically be implementing the same public functions as the `gkhash.c` file for `LMDB` 
... or better, for any key-value database.

Something about its `API` is mentioned in your documents [here](https://github.com/LMDB/lmdb/blob/LMDB_0.9.24/libraries/liblmdb/intro.doc)
which would be this [url](https://github.com/ahupowerdns/ahutils/blob/master/lmdb-semantics.md).
Examples are found in the source itself.

See you.

**Comment by allinurl** [comment]
@0bi-w6n-K3nobi I did test lmdb before replacing the previous on-disk version. It's possible that it may have needed some tuning, but still I think the performance wasn't there. Please see [**this post**](https://github.com/allinurl/goaccess/issues/1274#issuecomment-441272550), unfortunately not long ago I deleted the branch.

**Comment by 0bi-w6n-K3nobi** [comment]
Hi @allinurl .

Interesting... although `LMDB` have same evolution since your post.
It is a misfortune that has already erased.
There is an evolution: [libmdbx](https://github.com/erthink/libmdbx) or simply mdbx
with [`C API`](https://erthink.github.io/libmdbx/group__c__api.html) and automatic map resize.

Well... if you find old source/branch please share for turn so easy some tests.

See you next time.

**Comment by 0bi-w6n-K3nobi** [comment]
**Note:** Remembering here that it would be an alternative for those who have little memory [my case] or will process an absurd volume of requests.

**Comment by 0bi-w6n-K3nobi** [comment]
Well @allinurl .
I found in some mirror  at gitee [here](https://gitee.com/mirrors/goaccess/tree/lmdb).
But I think about replace `gkhash.c` public function, and not necessarily only storage.

**Comment by 0bi-w6n-K3nobi** [comment]
Good night, @allinurl .

See that you will post soon the new version 1.4.4. It's a great.
I tried understand the problem with JSON/HTML at patch f628472.
What was problem there?

So... I yet studing about a better key-value database.
In realy, both `lmdb` and `[lib]mdbx` have serius problems.
I'm using `ioarena` for tests. I had problems with `LevelDB` too.
I believe that `RocksDB`is great for our purpose.
But it is litte heavy, in code terms, but little memory comsuption that will be ideal for us.
I tested others,  through of `iorena`, but amost of them have a lot memory comsuption.

Well... On Monday I will test the new version 1.4.4 for large requests set,  but it will take a bit to give a return. 
[about 90 hours]

See you.

**Comment by allinurl** [comment]
@0bi-w6n-K3nobi Morning - I did crash on me when I raw a few tests. Still, you should be able to get to 4B without issues. As I look into the links you posted and options available, that specific function more likely will become a uint64_t. I'm running tests on different dbs as well and hope to have some news soon. Thanks again for the info. BTW, what's the file size of your log?

**Comment by 0bi-w6n-K3nobi** [comment]
Hi @allinurl.

In truly it is not only a `LOG`, but several` LOGs` during one month.
I have two options:

* November 2020 with 2,52 Bi requests and 38 TiB of Tx.
* December 2020 with 2,27 Bi requests and 32 TiB of Tx.

Both overflow 31bits for some counters, including Total Requests.
I will use these for testing of new 1.4.4 version.

Well... I liked very much of `IOWOW DB`. It's very simple to implement in code.
BUT, it have same memory consumption of `DB` size. It's very BAD for us.
You can just look for DBs into `IORENA` or another benchmark tools.
I just ignored `DBs` like `Sophia`, because it is without update.


See you next time.

**Comment by 0bi-w6n-K3nobi** [comment]
Hi @allinurl.

I found [this](https://github.com/pmwkaa/engine.so)... maybe it little bit old, but I believe work like reference.
And [this](https://ntnuopen.ntnu.no/ntnu-xmlui/bitstream/handle/11250/2506148/19718_FULLTEXT.pdf) interesting auto-tunning about `RocksDB`... Just a reference, you could read only if have some time for that.
See you.

**Comment by 0bi-w6n-K3nobi** [comment]
And more [here](https://db-engines.com/en/ranking/key-value+store).

**Comment by 0bi-w6n-K3nobi** [comment]
Hi @allinurl.

Finally, I have some **results** from **version 1.4.4** about `UINT64_T` counters.

Its seems to me that most things just right.
I observed a strange behavior about unique visitor.
In this new version the values decreased about roughly 0.75%.
Yeap. The `LOGS` and `total requests` are same.
You can see follow, compared with previous version:

![image](https://user-images.githubusercontent.com/71027865/106458312-f7b5d080-646e-11eb-94aa-cf180e2e03e9.png)![image](https://user-images.githubusercontent.com/71027865/106458573-59763a80-646f-11eb-94c4-3fc9c5188e65.png)

Well... In fact I changed a little bit the filtering, but only about `VHOST` names... removing `:80`, `:43`, `com`**`.`**, etc.
I verified in `VHOST` Panel and counters not change substantially to justify the change in visitors.

Coincidence or not, I found `GoAccess` a little slower.
The `Init. Proc. Time` counter was `23:43:49` vs `22:20:30` in the previous version. 
The time spend for running was 119h 45m.
I estimate that the previous time was 118h 30m.

And finally...
The sorting of data for `HTTP Status Codes` panel yet wrong.
Even thought the counters all positives, the last one behaves equally to the negative.
See follow, compared with previous version:

![image](https://user-images.githubusercontent.com/71027865/106461221-3cdc0180-6473-11eb-9fbc-92ab97ea3379.png)![image](https://user-images.githubusercontent.com/71027865/106461430-8593ba80-6473-11eb-8cbb-5f75b99a0b89.png)


And same for `The cache status of the object served` Panel:

![image](https://user-images.githubusercontent.com/71027865/106461686-de635300-6473-11eb-8720-54ee82074bde.png)![image](https://user-images.githubusercontent.com/71027865/106463908-bc1f0480-6476-11eb-8b69-acf14a4e4f28.png)

Well... I believe that this is all.

See You later.

**Comment by allinurl** [comment]
@0bi-w6n-K3nobi Thanks for posting these details!! I'll work on address this before the upcoming version.

Sorry for the delay on my reply, I've looking at different stores. BTW, so far LSM dbs are the most promising based on the tests that I have ran so far. Inserting billions of keys without really slowing down as opposed to btrees. Need to run some additional tests and I'll post back.

**Comment by 0bi-w6n-K3nobi** [comment]
Hi @allinurl. 

Yeah. The LSM technology is better for insert, while B-Tree for reading.
Until now I believe  that RocksDB is great.
Several projects in free software have had used for backend storage.
One of them is [Percona MyRocks](https://www.percona.com/doc/percona-server/8.0/myrocks/index.html) a implentation of MySql under RocksDB.
I can see a good review about performance [here](https://www.percona.com/doc/percona-server/8.0/myrocks/index.html).

But, for us the memory consumption is a important key.
If DB have a adjustable parameters for that, like cache, will be ideal.

What have you been thinking about the interface?
Mainly to replace the implementation of the `HASH` tables -- I.E `gkhash.c`. 
I believe that best solution should be keep current public functions and implement for new engine like DB, 
therefore this enable any new idea.
So, one solution will be create one executable for each engine implementation.

Well... this it only a draft.

**Comment by allinurl** [comment]
@0bi-w6n-K3nobi Good point. I think at some point we can have a separate implementation/executable taking care of the parsing and storing data and goaccess reading it. This will require to move (and possible maintain separately) the log parser into each implementation/executable, so more likely it won't be on the first iteration but perhaps a future one, especially since I still want to keep the current hash tables as the main engine.

**Comment by 0bi-w6n-K3nobi** [comment]
Hi @allinurl .

Just now that I see your post.
I think that it possible will be implement through of `MACRO`, that the strength point of Language C.

I was thinking here, with my buttons.
And I had a crazy idea, but not so original. 

**Begin Out of context note:** Remembering something, from my older brother.

In an old DOS game -- Prince of Persia -- when my brother ran it under virtualization and limiting memory to 128K,
using disk for page faults.
Good times, lousy hardware -- a 386SX computer.

**End Out of context note.**

Would it be possible implement `khash.h` in disk --  using `MACRO`s and replacing `malloc`'s functions -- ?
Using most likely `mmap` and large blocks like 2M for `malloc`'s functions replaced -- similar to `lmdb` database.
Thus, nothing would need to be changed. Just make use of `#include`s and `#define`s.

Just an idea ... If pagination becomes efficient, it would be a good implementation for those with little memory.

See you later.

**Comment by allinurl** [comment]
@0bi-w6n-K3nobi quick question, from your comment above:

> And finally...
> The sorting of data for HTTP Status Codes panel yet wrong.
> Even thought the counters all positives, the last one behaves equally to the negative.
> See follow, compared with previous version:
> 
> image1 image2
> 
> And same for The cache status of the object served Panel:

Are the images on the left-hand side the new version? From the images on the left, the counters appear to be positive > `uint32_t` so I'm curious if it's still occurring?

---

I do remember the old [prince of persia](https://www.retrogames.cz/games/102/DOS_01.gif), the movements upon jumping were awesome!

We're on the same page about having a version of khash on-disk. I thought about it too - I really like the idea of not using `(void*)` to achieve generic typing and that's what sold me. Using `void*` for strings is fine, but will cause overhead for integers, which are used widely in goaccess.

**Comment by 0bi-w6n-K3nobi** [comment]
Hi @allinurl .

Just yesterday afternoon I see your reply.
Yeap. Left-hand is new version... and problem with sort is same.

Well... **Good** and [I hope so] **Great news** for `khash`.
I tried to explain my tests and you will can draw your conclusions...

In my mind, I thinked that... `mmap`strategy is very similar to `swapping` pages for virtual memory.
Then, I needed simulate very few memory process/machine to force `swapping`, and keeping minimum into resident memory.
The Solution that I founded: **CGroup**.
You can read little "how to" [here](https://www.linuxjournal.com/content/everything-you-need-know-about-linux-containers-part-i-linux-control-groups-and-process).

So, my first tests into Centos 7.9 [kernel 3.10] were disastrous.
I needed set swapping memory in same value that resident memory... Intriguing!
Right after... I tested in my machine... Fedora 33 [Kernel 5.10], and results was great.
With around 512M, and rest into swap [not ssd], I got it running over first week of Febrary... 600 Mi requests... Great!
Recalling that, in this case, I used some simplifications:
* all `request lines` was replaced for `"GET / HTTP/1.1"`;
* enable only panels:  `VISITORS`, `OS`, `BROWSERS`, `VISIT_TIMES`, `VIRTUAL_HOSTS`, `STATUS_CODES`, `CACHE_STATUS` and `GEO_LOCATION`;

These conditions allow you to spend only 10 bytes, on average, per request.
In this case, around of 6 GiBytes.
In some times, the process had heavy `IO` from `swap`... mainly at end, for generate `HTML` output.
However, I believe it went well.

Well... So I needed to do a more realistic test.
To give you an idea ... the simplified `LOG` is 130Gi against complete ~???Gi~.
Again, I tried running .. and was so hard find good values/limits for resident memory... I was discouraged.

Well... So, I remembered the `ZRAM` device.
The first tests were confusing, so I remembered that `ZRAM` coexisted with `ZSWAP` ... 
Reminiscent of previous versions of Fedora which I upgraded.
So, I disabled `ZSWAP` and consulted [this](https://linuxreviews.org/Zram) little "how to" for `ZRAM`.

I yet running, processing the complete/full `LOG`.
Of course, a sufficient minimum of resident memory is required, avoiding unnecessary pressure on the `ZRAM`. 
For my case, I found value between 3/4 Gi a 1 Gi.
In short ... the great advantage in `ZRAM` is speed [of corse, is `RAM`] and compression: I found between 2.9 and 3.1.
In Other Words: I will spend only 1/3 of `RAM` in comparation with ordinal running, with litte bit slow down execution.


Well... Returning to idea about `nmap`.
Ok, I think that understand you about not using generic typing.
I did think about 1 `nmap` for each `HASH`... But it is possible for each type too, avoiding use of `(void *)`.
well ... I think while it's everything.

Unfortunately my Desk will not have memory to run all of 7 days, even with the use of `ZRAM`. 

See you.
